jax/interpreters/sharded_jit.py

```diff
@@ -300,3 +336,52 @@ def wrapped(*args, **kwargs):
     return tree_unflatten(out_tree(), out)
 
   return wrapped
+
+
+def _set_sharding_impl(x, partitions):
+  # TODO(skye): can we also prevent this from being called in other
+  # non-sharded_jit contexts? (e.g. pmap, control flow)
+  raise NotImplementedError(
+      "set_sharding() should only be called inside sharded_jit()")
+
+def _set_sharding_translation_rule(c, x_node, partitions):
+  return xb.set_sharding(c, x_node, partitions)
+
+set_sharding_p = core.Primitive("set_sharding")
+set_sharding_p.def_impl(_set_sharding_impl)
+set_sharding_p.def_abstract_eval(lambda x, partitions: x)
+ad.deflinear(set_sharding_p, lambda ct, partitions: (set_sharding(ct, partitions),))
+xla.translations[set_sharding_p] = _set_sharding_translation_rule
+
+def set_sharding(x, partitions: Optional[PartitionSpec]):
+  """Identity-like function that specifies how ``x`` should be sharded.
+
+  WARNING: this feature is still under active development! It may not work well,
+  and may change without warning!
+
+  This should only be called inside a function transformed by ``sharded_jit``.
+  It refines how ``f`` is sharded. ``partitions`` must correspond to the same
+  number of total partitions dictated by the outer ``sharded_jit`` and any other
+  ``set_sharding`` calls. In the case where only replication has been specified,
+  any ``partitions`` are valid.
+
+  Example usage:
+    @partial(sharded_jit, in_parts=None, out_parts=None, num_shards=2
+    def f(x):
+      y = x + 1
+      y = set_sharding(y, PartitionSpec(2,1))
+      return y * 2
+
+  In this example, the inputs and outputs of ``f`` will be replicated, but the
+  inner value of ``y`` will be partitioned in half. ``f`` will run on two
+  devices due to the set_sharding call.
+
+  Args:
+    x: Array value
+    partitions: PartitionSpec indicating how ``x`` should be partitioned, or
+      None for replication.
+
+  Returns:
+    A new version of ``x`` with the specified sharding applied.
```

Comment:
Is there a way to make it clear that this isn't a "please reshard this value from however it was sharded before to this new sharding" (even if maybe it could eventually mean that in a hypothetical op-by-op context) but more like "annotate this value as a constraint for the compiler, so that it's sharded in this way at runtime and the compiler will propagate that sharding forwards and/or backwards as it sees fit"?

Reply:
I added a bunch more verbiage in an attempt to explain this :) Lemme know if you have any suggestions to make it more clear.

Reply:
LGTM!